src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/AgreementsApiServiceImpl.scala

```diff
@@ -244,6 +253,44 @@ final case class AgreementsApiServiceImpl(
       eService                         <- catalogManagementService.getEService(agreement.eserviceId)
     } yield (producerTenant.name, consumerTenant.name, consumerTenant, eService)
 
+  def parallelGetTenantsService(agreement: AgreementProcess.Agreement)(implicit
+    contexts: Seq[(String, String)]
+  ): Future[(TenantManagement.Tenant, TenantManagement.Tenant, CatalogManagement.EService)] =
+    Applicative[Future].tuple3(
+      tenantManagementService
+        .getTenant(agreement.consumerId),
+      tenantManagementService.getTenant(agreement.producerId),
+      catalogManagementService.getEService(agreement.eserviceId)
+    )
+
+  def enrichAgreements(
+    agreement: AgreementProcess.Agreement
+  )(implicit contexts: Seq[(String, String)]): Future[CompactAgreements] = for {
+    (consumerTenant, producerTenant, eService) <- parallelGetTenantsService(agreement)
+  } yield CompactAgreements(
+    id = agreement.id,
+    consumer = CompactOrganization(consumerTenant.id, consumerTenant.name),
+    eservice = CompactEService(
+      id = eService.id,
+      name = eService.name,
+      producer = CompactOrganization(producerTenant.id, producerTenant.name)
+    ),
+    upgradable = isUpgradable(agreement.descriptorId.toString, eService.descriptors).getOrElse(false)
+  )
+
+  def isUpgradable(descriptorId: String, descriptors: Seq[CatalogManagement.EServiceDescriptor]): Option[Boolean] =
+    for {
+      currentDescriptor <- descriptors.find(_.id == descriptorId).headOption
+      upgradable        <- descriptors
+        .filter(_.version.toInt > currentDescriptor.version.toInt)
+        .find(d =>
+          d.state == CatalogManagement.EServiceDescriptorState.PUBLISHED ||
+            d.state == CatalogManagement.EServiceDescriptorState.SUSPENDED
+        )
+        .headOption
+
+    } yield upgradable.some.isDefined
```

Comment:
no need to return an `Option` in this function
```suggestion
      upgradable        = descriptors
        .filter(_.version.toInt > currentDescriptor.version.toInt)
        .find(d =>
          d.state == CatalogManagement.EServiceDescriptorState.PUBLISHED ||
            d.state == CatalogManagement.EServiceDescriptorState.SUSPENDED
        )
    } yield upgradable.isDefined
```